### PR TITLE
Add bone noise

### PR DESCRIPTION
**Pull request by mac2648**
The bone aim is quite bad now, but I will change whenever we have the first person perspective.
For the bone to trigger the AI hearing we need to set the ground and walls to generate overlap events, I will search later if there is a better way to do it that does not require to add overlap events to all ground and walls.
Changed the bone root component so the box component follows the static mesh

### CODE REVIEW

File: Source/Redemption/Private/Bone.cpp [hunk 1]
```diff
@@ -10,12 +12,14 @@ ABone::ABone()
 	PrimaryActorTick.bCanEverTick = true;
 
 	BoxComp = CreateDefaultSubobject<UBoxComponent>(TEXT("Box Collider"));
-	SetRootComponent(BoxComp);
 
 	BoxComp->OnComponentBeginOverlap.AddDynamic(this, &ABone::OnBoneBeginOverlap);
 
 	BoneMesh = CreateDefaultSubobject<UStaticMeshComponent>(TEXT("Bone Mesh"));
-	BoneMesh->SetupAttachment(BoxComp);
```
Comment (Entwistle-905) [inline Source/Redemption/Private/Bone.cpp:18]: Why did you change this? Because I was mostly told to put box component (box collider) to be root component. is there any reason?

File: Source/Redemption/Private/PowerUpComponent.cpp [hunk 2]
```diff
@@ -83,4 +87,24 @@ void UPowerUpComponent::ChangeActivePowerUp(EActivePowerUp NewPower)
...
+		{
+			FVector FowardVec = OwnerChar->GetActorForwardVector();
+			FVector SpawnPos = OwnerChar->GetActorLocation() + FowardVec * 100;
+
+			FRotator CharacterRotation = FowardVec.Rotation();
+
+			FVector BoneDirection = FowardVec + FVector(0.0, 0.0, 0.3);
+
+			ABone* Bone = GetWorld()->SpawnActor<ABone>(BoneBP, SpawnPos, FRotator::ZeroRotator);
+			Bone->GetMesh()->AddImpulse(BoneDirection * BONE_SPEED);
```
Comment (Entwistle-905) [inline Source/Redemption/Private/PowerUpComponent.cpp:107]: In the comment, you said that this code will spawn a bone, but it seems like throwing a bone. If this is about throwing in a bone in front of the player, we can't set the destination of the bone right?

File: Source/Redemption/Private/Bone.cpp (hunk 1 above)
Comment (mac2648) [reply Source/Redemption/Private/Bone.cpp:18]: When the bone moves what is moving is the static mesh. it is receiving the impulse. If the root is the box component it will not follow the mesh.
I changed it so the box component follows the mesh. 
But you are right in the point that in most cases it will be the box component to be the root

File: Source/Redemption/Private/PowerUpComponent.cpp (hunk 2 above)
Comment (mac2648) [reply Source/Redemption/Private/PowerUpComponent.cpp:107]: yes you are right it spawns and throws the bone. at the moment the bone will always move in front of the player. I will change that when we have the first perspective

File: Source/Redemption/Private/Bone.cpp (hunk 1 above)
Comment (Entwistle-905) [reply Source/Redemption/Private/Bone.cpp:18]: Aha thank you so much^-^

File: Source/Redemption/Private/Bone.cpp [hunk 3]
```diff
@@ -32,7 +36,20 @@ void ABone::Tick(float DeltaTime)
...
+//this function is triggered when the bone overlaps with something
 void ABone::OnBoneBeginOverlap(UPrimitiveComponent* OverlappedComp, AActor* OtherActor, UPrimitiveComponent* OtherComp, int32 OtherBodyIndex, bool bFromSweep, const FHitResult& SweepResult)
 {
-	
+	if (OtherActor == this)
+	{
+		return;
+	}
+
+	if (OtherActor != UGameplayStatics::GetPlayerPawn(GetWorld(), 0))
```
Comment (Fedic) [inline Source/Redemption/Private/Bone.cpp:47]: Is there a chance that OtherActor == nullptr?

File: Source/Redemption/Private/PowerUpComponent.cpp [hunk 4]
```diff
@@ -83,4 +87,24 @@ void UPowerUpComponent::ChangeActivePowerUp(EActivePowerUp NewPower)
...
+	if (BoneBP)
+	{
+		if (ACharacter* OwnerChar = Cast<ACharacter>(GetOwner()))
+		{
+			FVector FowardVec = OwnerChar->GetActorForwardVector();
+			FVector SpawnPos = OwnerChar->GetActorLocation() + FowardVec * 100;
+
+			FRotator CharacterRotation = FowardVec.Rotation();
+
+			FVector BoneDirection = FowardVec + FVector(0.0, 0.0, 0.3);
```
Comment (Fedic) [inline Source/Redemption/Private/PowerUpComponent.cpp:104]: Can this be a UProperty so we can change it in Unreal, so we can adjust how far the character throws it easier?

File: Source/Redemption/Private/PowerUpComponent.cpp [hunk 5]
```diff
@@ -83,4 +87,24 @@ void UPowerUpComponent::ChangeActivePowerUp(EActivePowerUp NewPower)
...
+
+			ABone* Bone = GetWorld()->SpawnActor<ABone>(BoneBP, SpawnPos, FRotator::ZeroRotator);
+			Bone->GetMesh()->AddImpulse(BoneDirection * BONE_SPEED);
+		}
+	}
 }
```
Comment (Fedic) [inline Source/Redemption/Private/PowerUpComponent.cpp:110]: Can we add something to delete the bone from the world after some amount of time for optimization purposes?

**Review by Fedic — CHANGES_REQUESTED** [review]

File: Source/Redemption/Private/PowerUpComponent.cpp (hunk 2 above)
Comment (Entwistle-905) [reply Source/Redemption/Private/PowerUpComponent.cpp:107]: Okay, thank you. I can't wait to see that

File: Source/Redemption/Private/Bone.cpp (hunk 3 above)
Comment (mac2648) [reply Source/Redemption/Private/Bone.cpp:47]: it has to hit something, so i don't think there is. I never saw it being nullptr

File: Source/Redemption/Private/PowerUpComponent.cpp (hunk 5 above)
Comment (mac2648) [reply Source/Redemption/Private/PowerUpComponent.cpp:110]: it will be deleted after it hits something

Comment (mac2648) [reply Source/Redemption/Private/PowerUpComponent.cpp:110]: if you want I can put a timer to delete it even if it does not hit anything

File: Source/Redemption/Private/PowerUpComponent.cpp (hunk 4 above)
Comment (mac2648) [reply Source/Redemption/Private/PowerUpComponent.cpp:104]: I will change that when we have the first person perspective, right now it is like this just for me to test the trigger

**Review by Entwistle-905 — APPROVED** [review]
